solver/src/liquidity/offchain_orderbook.rs

```diff
@@ -47,6 +51,194 @@ pub fn normalize_limit_order(order: Order, native_token: WETH9) -> LimitOrder {
 
 impl SettlementHandling<LimitOrder> for OrderSettlementHandler {
     fn encode(&self, executed_amount: U256, encoder: &mut SettlementEncoder) -> Result<()> {
-        encoder.add_trade(self.order.clone(), executed_amount)
+        if self.order.order_creation.buy_token == BUY_ETH_ADDRESS {
+            let interaction = compute_unwrap_interaction(
+                encoder.clearing_prices(),
+                &self.order,
+                executed_amount,
+                self.native_token.clone(),
+            )?;
+
+            encoder.add_trade(self.order.clone(), executed_amount)?;
+
+            encoder.add_unwrap(interaction);
+        } else {
+            encoder.add_trade(self.order.clone(), executed_amount)?;
+        }
```

Comment:
nit: Why not factor out the common `add_trade`?
```suggestion
            encoder.add_unwrap(interaction);
        }
        encoder.add_trade(self.order.clone(), executed_amount)?;
```

Reply:
The repetition is there because I wanted to guarantee that if the result is `Err` then no changes have been made to `encoder` (under the assumption that if `add_trade` returns `Err`, then the same is true).
With the proposed ordering, if `add_trade` fails then there could still be a new interaction added to `encoder`.

Reply:
Ah, that makes perfect sense.